Src/FluentAssertions/Equivalency/DataRowCollectionEquivalencyStep.cs

```diff
@@ -0,0 +1,268 @@
+﻿using System;
+using System.Data;
+using System.Linq;
+
+using FluentAssertions.Data;
+using FluentAssertions.Execution;
+
+namespace FluentAssertions.Equivalency
+{
+    public class DataRowCollectionEquivalencyStep : IEquivalencyStep
+    {
+        public bool CanHandle(IEquivalencyValidationContext context, IEquivalencyAssertionOptions config)
+        {
+            return typeof(DataRowCollection).IsAssignableFrom(config.GetExpectationType(context.RuntimeType, context.CompileTimeType));
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Style", "IDE0038:Use pattern matching", Justification = "Would decrease code clarity")]
+        public bool Handle(IEquivalencyValidationContext context, IEquivalencyValidator parent, IEquivalencyAssertionOptions config)
+        {
+            if (!(context.Subject is DataRowCollection))
+            {
+                AssertionScope.Current
+                    .FailWith("Expected {context:value} to be of type DataRowCollection, but found {0}", context.Subject.GetType());
+            }
+            else
+            {
+                RowMatchMode rowMatchMode = RowMatchMode.Index;
+
+                if (config is DataEquivalencyAssertionOptions<DataSet> dataSetConfig)
+                {
+                    rowMatchMode = dataSetConfig.RowMatchMode;
+                }
+                else if (config is DataEquivalencyAssertionOptions<DataTable> dataTableConfig)
+                {
+                    rowMatchMode = dataTableConfig.RowMatchMode;
+                }
+
+                var subject = (DataRowCollection)context.Subject;
+                var expectation = (DataRowCollection)context.Expectation;
+
+                AssertionScope.Current
+                    .ForCondition(subject.Count == expectation.Count)
+                    .FailWith("Expected {context:DataRowCollection} to contain {0} row(s){reason}, but found {1}", expectation.Count, subject.Count);
+
+                switch (rowMatchMode)
+                {
+                    case RowMatchMode.Index:
+                    {
```

Comment:
We haven't introduced an extra scope elsewhere for `case` statements.
(It also triggers an overzealous style Visual Studio style warning for me)